File: src/main/java/project/kconnecta/user/backend/feature/friend/service/FriendService.java
package project.kconnecta.user.backend.feature.friend.service;

import project.kconnecta.user.backend.feature.friend.dto.response.FriendBirthdayResponse;
import project.kconnecta.user.backend.feature.friend.dto.response.FriendResponse;
import project.kconnecta.user.backend.feature.friend.dto.response.FriendshipStatusResponse;

import java.util.List;
import java.util.UUID;

public interface FriendService {
    List<FriendResponse> getFriends(UUID userId);
    List<FriendBirthdayResponse> getFriendBirthdays(UUID userId);
    List<FriendResponse> getFriendRequests(UUID userId);
    List<FriendResponse> getSentFriendRequests(UUID userId);
    List<FriendResponse> getSuggestions(UUID userId);
    FriendResponse sendFriendRequest(UUID requesterId, UUID addresseeId);
    FriendResponse acceptFriendRequest(UUID friendshipId);
    void deleteFriendship(UUID friendshipId, UUID currentUserId);
    FriendshipStatusResponse getStatus(UUID meId, UUID targetId);
}

File: src/services/friendService.ts
import { api } from './api';

export const FRIENDSHIP_CHANGED_EVENT = 'friendship-changed';

export interface FriendApiResponse {
  friendshipId: string | null;
  userId: string;
  username: string;
  fullName: string;
  avatarUrl: string | null;
  mutualFriends: number;
  suggestionReason?: string | null;
  status: 'PENDING' | 'ACCEPTED' | null;
  createdAt: string | null;
}

export interface FriendshipStatusResponse {
  friendshipId: string | null;
  status: 'PENDING' | 'ACCEPTED' | null;
  sentByMe: boolean;
}

function notifyFriendshipChanged() {
  window.dispatchEvent(new Event(FRIENDSHIP_CHANGED_EVENT));
}

export const friendService = {
  getFriends: (userId: string) =>
    api.get<FriendApiResponse[]>(`/friends/${userId}`),

  getFriendRequests: () =>
    api.get<FriendApiResponse[]>('/friends/requests'),

  getSentFriendRequests: () =>
    api.get<FriendApiResponse[]>('/friends/sent-requests'),

  getSuggestions: (userId: string) =>
    api.get<FriendApiResponse[]>(`/friends/${userId}/suggestions`),

  sendFriendRequest: (requesterId: string, addresseeId: string) => {
    if (requesterId === addresseeId) {
      return Promise.reject(new Error('Không thể gửi lời mời kết bạn cho chính mình'));
    }
    return api.post<FriendApiResponse>('/friends/request', { requesterId, addresseeId });
  },

  acceptFriendRequest: async (friendshipId: string) => {
    const res = await api.put<FriendApiResponse>(`/friends/${friendshipId}/accept`, {});
    notifyFriendshipChanged();
    return res;
  },

  deleteFriendship: async (friendshipId: string) => {
    await api.delete<void>(`/friends/${friendshipId}`);
    notifyFriendshipChanged();
  },

  getStatus: (meId: string, targetId: string) =>
    api.get<FriendshipStatusResponse>(`/friends/status?me=${meId}&target=${targetId}`),
};
